File: extract-token.py
#!/usr/bin/env python3
import browser_cookie3
import sys

try:
    cj = browser_cookie3.chrome(domain_name='udemy.com')
    cookies = {cookie.name: cookie.value for cookie in cj}
    
    access_token = cookies.get('access_token', '').strip('"')
    client_id = cookies.get('client_id', '')
    
    if not access_token or not client_id:
        print("Error: Could not find access_token or client_id in Chrome cookies.", file=sys.stderr)
        print("Make sure you're logged into Udemy in Chrome.", file=sys.stderr)
        sys.exit(1)
    
    if '--json' in sys.argv:
        import json
        print(json.dumps({'access_token': access_token, 'client_id': client_id}))
    else:
        print(f"Access Token: {access_token}")
        print(f"Client ID: {client_id}")
except Exception as e:
    print(f"Error: {e}", file=sys.stderr)
    sys.exit(1)

File: extract-token.js
#!/usr/bin/env node
const { execSync } = require('child_process');
const path = require('path');

function extractCookies() {
  try {
    const scriptPath = path.join(__dirname, 'extract-token.py');
    const result = execSync(`python3 "${scriptPath}" --json`, { encoding: 'utf8' });
    return JSON.parse(result.trim());
  } catch (error) {
    return null;
  }
}

function autoLoadToken(Settings) {
  if (!Settings.accessToken) {
    const cookies = extractCookies();
    if (cookies) {
      Settings.accessToken = cookies.access_token;
      console.log('Auto-loaded access token from Chrome cookies');
      return true;
    }
  }
  return false;
}

if (require.main === module) {
  try {
    const cookies = extractCookies();
    if (cookies) {
      console.log('Access Token:', cookies.access_token);
      console.log('Client ID:', cookies.client_id);
    } else {
      console.error('Failed to extract cookies. Make sure you are logged into Udemy in Chrome.');
      process.exit(1);
    }
  } catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
  }
}

module.exports = { extractCookies, autoLoadToken };
